crates/exex/src/manager.rs

```diff
@@ -0,0 +1,396 @@
+use std::{
+    collections::VecDeque,
+    future::Future,
+    pin::Pin,
+    sync::{
+        atomic::{AtomicUsize, Ordering},
+        Arc,
+    },
+    task::{Context, Poll},
+};
+
+use crate::ExExEvent;
+use metrics::Gauge;
+use reth_metrics::{metrics::Counter, Metrics};
+use reth_primitives::BlockNumber;
+use reth_provider::CanonStateNotification;
+use tokio::sync::{
+    mpsc::{self, error::SendError, Receiver, UnboundedReceiver, UnboundedSender},
+    watch,
+};
+use tokio_util::sync::{PollSendError, PollSender};
+
+/// Metrics for an ExEx.
+#[derive(Metrics)]
+#[metrics(scope = "exex")]
+struct ExExMetrics {
+    /// The number of canonical state notifications processed by an ExEx.
+    notifications_processed: Counter,
```

Comment:
shouldn't it rather be "sent" because we don't know if ExEx processed any?
```suggestion
    /// The number of canonical state notifications sent to an ExEx.
    notifications_sent: Counter,
```

Reply:
also counter naming convention
```suggestion
    /// The number of canonical state notifications processed by an ExEx.
    notifications_sent_total: Counter,
```

Reply:
we assumed sent = processed, the exex should not assume it will receive the notification again